File: backend/app/services/scheduler.py
"""
Background scheduler that fetches flight data from AirLabs every 10 minutes,
persists it to flights_cache.json, and broadcasts an SSE refresh event to all
connected browser clients.

Usage (in main.py lifespan):

    from app.services.scheduler import FlightScheduler
    scheduler = FlightScheduler(airlabs_service, settings, sse_broadcaster)
    task = asyncio.create_task(scheduler.run())
    ...
    task.cancel()
"""

import asyncio
import datetime
import logging
from pathlib import Path
from typing import TYPE_CHECKING

from app.utils.json_store import write_json

if TYPE_CHECKING:
    from app.services.airlabs import AirLabsService
    from app.config import Settings

logger = logging.getLogger(__name__)


class SSEBroadcaster:
    """Manages a set of per-client asyncio Queues for Server-Sent Events."""

    def __init__(self) -> None:
        self._clients: set[asyncio.Queue] = set()
        self._lock = asyncio.Lock()

    async def subscribe(self) -> asyncio.Queue:
        q: asyncio.Queue = asyncio.Queue(maxsize=10)
        async with self._lock:
            self._clients.add(q)
        return q

    async def unsubscribe(self, q: asyncio.Queue) -> None:
        async with self._lock:
            self._clients.discard(q)

    async def broadcast(self, message: str) -> None:
        """Push *message* to every connected client queue (non-blocking)."""
        async with self._lock:
            dead: set[asyncio.Queue] = set()
            for q in self._clients:
                try:
                    q.put_nowait(message)
                except asyncio.QueueFull:
                    dead.add(q)
            self._clients -= dead
        if dead:
            logger.debug("SSE: dropped %d slow/full client queue(s)", len(dead))

    @property
    def client_count(self) -> int:
        return len(self._clients)


class FlightScheduler:
    """Periodically fetches default-region flights and writes to JSON."""

    def __init__(
        self,
        airlabs_service: "AirLabsService",
        cache_path: Path,
        interval_seconds: int,
        broadcaster: SSEBroadcaster,
    ) -> None:
        self._service = airlabs_service
        self._cache_path = cache_path
        self._interval = interval_seconds
        self._broadcaster = broadcaster

    async def run(self) -> None:
        logger.info(
            "FlightScheduler started — interval=%ds, cache=%s",
            self._interval,
            self._cache_path,
        )
        while True:
            await self._tick()
            await asyncio.sleep(self._interval)

    async def _tick(self) -> None:
        fetched_at = datetime.datetime.utcnow().isoformat() + "Z"
        logger.info("FlightScheduler: fetching default-region flights…")
        try:
            response = await self._service.get_default_flights()
            payload = {
                "fetched_at": fetched_at,
                "count": response.count,
                "bbox": list(response.bbox),
                "source_time": response.source_time,
                "flights": [f.model_dump() for f in response.flights],
            }
            await write_json(self._cache_path, payload)
            logger.info(
                "FlightScheduler: wrote %d flights to %s",
                response.count,
                self._cache_path,
            )

            # Broadcast SSE refresh event to all connected browsers
            import json
            msg = json.dumps(
                {
                    "type": "refresh",
                    "count": response.count,
                    "fetched_at": fetched_at,
                },
                separators=(",", ":"),
            )
            await self._broadcaster.broadcast(f"data: {msg}\n\n")

        except Exception as exc:
            logger.error("FlightScheduler tick failed: %s", exc, exc_info=True)

File: backend-node/src/services/scheduler.js
// ── services/scheduler.js ─────────────────────────────────────────────────────
// Background flight scheduler + SSE broadcaster.
// Port of Python backend/app/services/scheduler.py
// ──────────────────────────────────────────────────────────────────────────────

import { writeJson } from '../utils/jsonStore.js';
import { appendFileSync } from 'fs';

// ── SSEBroadcaster ────────────────────────────────────────────────────────────
// Manages a set of active SSE response objects. When the scheduler fetches new
// flight data it calls broadcast() to push the event to every open connection.
// ─────────────────────────────────────────────────────────────────────────────

export class SSEBroadcaster {
  constructor() {
    /** @type {Set<import('express').Response>} */
    this._clients = new Set();
    this.onClientCountChange = null;
  }

  get clientCount() { return this._clients.size; }

  /**
   * Register a new SSE client (the Express `res` object).
   * @param {import('express').Response} res
   */
  subscribe(res) {
    this._clients.add(res);
    if (typeof this.onClientCountChange === 'function') {
      this.onClientCountChange(this.clientCount);
    }
  }

  /**
   * Remove an SSE client when its connection closes.
   * @param {import('express').Response} res
   */
  unsubscribe(res) {
    this._clients.delete(res);
    if (typeof this.onClientCountChange === 'function') {
      this.onClientCountChange(this.clientCount);
    }
  }

  /**
   * Push `message` (a fully-formatted SSE string, e.g. "data: {...}\n\n")
   * to every connected client. Dead connections are cleaned up automatically.
   * @param {string} message
   */
  broadcast(message) {
    const dead = [];
    for (const res of this._clients) {
      try {
        res.write(message);
      } catch {
        dead.push(res);
      }
    }
    for (const res of dead) this._clients.delete(res);
    if (dead.length > 0) {
      console.debug(`[SSEBroadcaster] dropped ${dead.length} closed connection(s)`);
      if (typeof this.onClientCountChange === 'function') {
        this.onClientCountChange(this.clientCount);
      }
    }
  }
}

// ── FlightScheduler ───────────────────────────────────────────────────────────
// Periodically fetches default-region flights, writes them to JSON cache, and
// broadcasts an SSE refresh event to all connected browsers.
// ─────────────────────────────────────────────────────────────────────────────

export class FlightScheduler {
  /**
   * @param {{
   *   airlabsService: import('./airlabs.js').AirLabsService,
   *   cachePath: string,
   *   intervalSeconds: number,
   *   broadcaster: SSEBroadcaster,
   * }} opts
   */
  constructor({ airlabsService, cachePath, intervalSeconds, broadcaster }) {
    this._service     = airlabsService;
    this._cachePath   = cachePath;
    this._intervalMs  = intervalSeconds * 1000;
    this._broadcaster = broadcaster;
    this._timer       = null;
    this._idleTimeout = null;
  }

  /** Start the scheduler: run immediately, then every intervalMs. */
  start() {
    console.info(
      `[FlightScheduler] started — interval=${this._intervalMs / 1000}s, cache=${this._cachePath}`,
    );

    // Register callback to react to SSE connection events
    this._broadcaster.onClientCountChange = (count) => this._handleClientCountChange(count);

    this._tick();                                  // immediate first run
    this._timer = setInterval(() => this._tick(), this._intervalMs);

    // Assess initial inactivity timeout setup (since client count starts at 0)
    this._handleClientCountChange(this._broadcaster.clientCount);
  }

  /** Stop the scheduler (call on server shutdown). */
  stop() {
    if (this._timer) {
      clearInterval(this._timer);
      this._timer = null;
    }
    if (this._idleTimeout) {
      clearTimeout(this._idleTimeout);
      this._idleTimeout = null;
    }
  }

  _handleClientCountChange(count) {
    if (count > 0) {
      if (this._idleTimeout) {
        console.info('[FlightScheduler] Activity detected. Cancelling scheduled suspend.');
        clearTimeout(this._idleTimeout);
        this._idleTimeout = null;
      }
      if (!this._timer) {
        this._resume();
      }
    } else {
      // count === 0
      if (this._timer && !this._idleTimeout) {
        console.info('[FlightScheduler] 0 clients connected. Scheduling suspend in 10 minutes.');
        this._idleTimeout = setTimeout(() => {
          this._pause();
        }, 10 * 60 * 1000); // 10 minutes
      }
    }
  }

  _pause() {
    console.info('[FlightScheduler] Pausing flight scheduler due to inactivity (0 clients for >10m).');
    if (this._timer) {
      clearInterval(this._timer);
      this._timer = null;
    }
    this._idleTimeout = null;
  }

  _resume() {
    console.info('[FlightScheduler] Active clients detected. Resuming flight scheduler.');
    this._tick();
    if (this._timer) clearInterval(this._timer);
    this._timer = setInterval(() => this._tick(), this._intervalMs);
  }

  async _tick() {
    const fetchedAt = new Date().toISOString();
    console.info('[FlightScheduler] fetching default-region flights…');
    try {
      const response = await this._service.getDefaultFlights();
      
      console.log("Flights Parsed:", response.flights.length);
      if (response.flights.length > 0) {
        console.log("Sample Flight:", response.flights[0]);
      }

      const schedulerLog = `\n--- SCHEDULER DIAGNOSTICS ---\nTimestamp: ${new Date().toISOString()}\nFlights Parsed: ${response.flights.length}\nSample Flight: ${response.flights.length > 0 ? JSON.stringify(response.flights[0]) : 'None'}\n-----------------------------\n`;
      try { appendFileSync('v:\\BUP\\backend-node\\data\\diagnostics.log', schedulerLog); } catch (e) {}

      const payload = {
        fetched_at:  fetchedAt,
        count:       response.count,
        bbox:        response.bbox,
        source_time: response.source_time,
        flights:     response.flights,
      };
      await writeJson(this._cachePath, payload);
      console.info(`[FlightScheduler] wrote ${response.count} flights to ${this._cachePath}`);

      // Broadcast SSE refresh to all connected browsers
      const msg = JSON.stringify({ type: 'refresh', count: response.count, fetched_at: fetchedAt });
      this._broadcaster.broadcast(`data: ${msg}\n\n`);
    } catch (err) {
      console.error(`[FlightScheduler] tick failed: ${err.message}`, err);
    }
  }
}
